File: csharp/need-for-speed/NeedForSpeed.cs
using System;

class RemoteControlCar
{
    public int speed;
    public int batteryDrain;
    private int distanceDriven = 0;
    private int battery = 100;

    public RemoteControlCar(int speed, int batteryDrain)
    {
        this.speed = speed;
        this.batteryDrain = batteryDrain;
    }

    public bool BatteryDrained()
    {
        if (battery <= 0 | battery < this.batteryDrain)
        {
            return true;
        }
        return false;
    }

    public int DistanceDriven()
    {
        return distanceDriven;
    }

    public void Drive()
    {
        if (!BatteryDrained())
        {
            distanceDriven += this.speed;
            battery -= this.batteryDrain;
        }
    }

    public static RemoteControlCar Nitro()
    {
        var car = new RemoteControlCar(50, 4);
        return car;
    }
}

class RaceTrack
{
    private float distance;

    public RaceTrack(int distance)
    {
        this.distance = distance;
    }

    public bool TryFinishTrack(RemoteControlCar car)
    {
        float autonomy = (float)car.speed / (float)car.batteryDrain;

        if (this.distance / 100 <= autonomy)
        {
            return true;
        }
        else
        {
            return false;
        }

    }
}

File: java/need-for-speed/src/main/java/NeedForSpeed.java
class NeedForSpeed {

    private int speed = 5;
    private int batteryDrain = 0;
    private int distanceDriven = 0;
    private int remainingBattery = 100;

    public NeedForSpeed(int speed, int batteryDrain) {
        this.speed = speed;
        this.batteryDrain = batteryDrain;
    }

    public boolean batteryDrained() {
        if (this.remainingBattery < this.batteryDrain) {
            return true;
        }
        return false;
    }

    public int distanceDriven() {
        return this.distanceDriven;
    }

    public void drive() {
        if (!batteryDrained()) {
            this.distanceDriven = this.distanceDriven + this.speed;
            this.remainingBattery = this.remainingBattery - this.batteryDrain;
        }
    }

    public static NeedForSpeed nitro() {
        NeedForSpeed car = new NeedForSpeed(50, 4);
        return car;
    }
}

class RaceTrack {

    private int distance;

    public RaceTrack(int distance) {
        this.distance = distance;
    }

    public boolean tryFinishTrack(NeedForSpeed car) {
        while (!car.batteryDrained()) {
            car.drive();
        }

        if (car.distanceDriven() >= this.distance) {
            return true;
        }
        return false;
    }
}
